velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/prestosql/types/JsonType.h"
+
+#include "folly/Conv.h"
+#include "folly/json.h"
+
+#include "velox/functions/lib/StringEncodingUtils.h"
+
+namespace facebook::velox {
+
+bool JsonCastOperator::isSupportedType(const TypePtr& other) const {
+  switch (other->kind()) {
+    case TypeKind::BOOLEAN:
+    case TypeKind::BIGINT:
+    case TypeKind::INTEGER:
+    case TypeKind::SMALLINT:
+    case TypeKind::TINYINT:
+    case TypeKind::DOUBLE:
+    case TypeKind::REAL:
+    case TypeKind::VARCHAR:
+    case TypeKind::DATE:
+    case TypeKind::TIMESTAMP:
+      return true;
+    case TypeKind::ARRAY:
+      return isSupportedType(other->childAt(0));
+    case TypeKind::ROW:
+      for (auto& child : other->as<TypeKind::ROW>().children()) {
+        if (!isSupportedType(child)) {
+          return false;
+        }
+      }
+      return true;
+    case TypeKind::MAP:
+      return (
+          other->childAt(0)->kind() == TypeKind::VARCHAR &&
+          isSupportedType(other->childAt(1)));
+    default:
+      return false;
+  }
+}
+
+void JsonCastOperator::castTo(
+    const SelectivityVector& rows,
+    exec::EvalCtx* context,
+    const VectorPtr& input,
+    DecodedVector* inputDecoded,
+    const TypePtr& fromType,
+    VectorPtr* result) {
+  VectorPtr localResult;
+  VectorPtr emptyVectorPtr;
+  functions::prepareFlatResultsVector(
+      &localResult, rows, context, emptyVectorPtr);
+  auto* flatResult = localResult->as<FlatVector<StringView>>();
+
+  exec::LocalSelectivityVector baseRows(
+      context->execCtx(), inputDecoded->base()->size());
+  baseRows->clearAll();
+  rows.applyToSelected(
+      [&](auto row) { baseRows->setValid(inputDecoded->index(row), true); });
+  baseRows->updateBounds();
+
+  baseRows->applyToSelected([&](auto row) {
+    result_.clear();
```

Comment:
It is not clear what are the benefits of using result_ member variable. Since result_ is an std::string, we end up allocating memory and copying it for each row. The code would be clearer if generateJson simply returned std::string. 

To avoid extra memory allocation and copy, we'll need to modify generateJson to use FlatVector::getBufferWithSpace (to make sure there is enough space in the vector's buffer) and write directly to that buffer.